lib/OptimizingHandler.js

```diff
@@ -0,0 +1,112 @@
+'use strict';
+
+const SparqlGenerator = require( 'sparqljs' ).Generator;
+const traverse = require( 'traverse' );
+const { allPrefixes } = require( './rdfNamespaces' );
+const WdqsRequest = require( './WdqsRequest' );
+
+module.exports = class OptimizingHandler {
+
+	constructor() {
+		this.generator = new SparqlGenerator( { prefixes: allPrefixes } );
+	}
+
+	handle( query, parsedQuery ) {
+		if ( !parsedQuery ||
+			parsedQuery.queryType !== 'SELECT'
+		) {
+			return;
+		}
+
+		let hasChange = false;
+		/* eslint-disable array-callback-return */ // false positive
+		const newQuery = traverse( parsedQuery ).map( function ( element ) {
+			if ( element.type !== 'filter' ||
+				element.expression.type !== 'operation' ||
+				element.expression.operator !== '=' ||
+				element.expression.args.length !== 2
+			) {
+				return;
+			}
+			let [ lhs, rhs ] = element.expression.args;
+			if ( lhs.type !== 'operation' ) {
+				// try swapping them
+				[ lhs, rhs ] = [ rhs, lhs ];
+			}
+			if ( lhs.type !== 'operation' ||
+				lhs.operator !== 'year' ||
+				lhs.args.length !== 1 ||
+				lhs.args[ 0 ].termType !== 'Variable' ||
+				rhs.termType !== 'Literal' ||
+				rhs.datatype.value !== 'http://www.w3.org/2001/XMLSchema#integer'
+			) {
+				return;
+			}
+
+			let year;
+			try {
+				year = parseInt( rhs.value );
+			} catch ( e ) {
+				return;
+			}
+			if ( year.toString() !== rhs.value ) {
+				// precision issues or garbage at end of input
+				return;
+			}
+			if ( year === year + 1 ) {
+				// precision issues
+				return;
+			}
+
+			// we actually have an expression of the form FILTER(YEAR(?date) = 2020); rewrite it to
+			// FILTER(?date >= "2020-00-00"^^xsd:dateTime && ?date < "2021-00-00"^^xsd:dateTime)
```

Comment:
Cool. Comments like this are definitely necessary when hand-crafting expressions like this in the AST.